db/migrate/20150716133712_add_unique_index_on_journals.rb

```diff
@@ -0,0 +1,171 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddUniqueIndexOnJournals < ActiveRecord::Migration
+  def up
+    cleanup_duplicate_journals
+    add_index :journals, [:journable_type, :journable_id, :version], unique: true
+  end
+
+  def down
+    remove_index :journals, [:journable_type, :journable_id, :version]
+  end
+
+  private
+
+  def cleanup_duplicate_journals
+    duplicate_pairs = find_duplicate_journal_ids
+
+    if duplicate_pairs.any?
+      say "Found #{duplicate_pairs.count} journals with at least one duplicate!"
+      say_with_time 'Safely removing duplicates...' do
+        undeleted_pairs = []
+        duplicate_pairs.each do |current_id, duplicate_id|
+          say "Comparing journals ##{current_id} & ##{duplicate_id} for equality", subitem: true
+
+          current = MigrationHelperJournal.find(current_id)
+          duplicate = MigrationHelperJournal.find(duplicate_id)
+
+          if journals_equivalent?(current, duplicate)
+            say "Deleting journal ##{current.id}...", subitem: true
+            current.destroy
+          else
+            undeleted_pairs << [current_id, duplicate_id]
+          end
+        end
+
+        abort_on_undeleted_pairs undeleted_pairs
+      end
+    end
+  end
+
+  def find_duplicate_journal_ids
+    this = MigrationHelperJournal.table_name
+    MigrationHelperJournal
+      .joins("INNER JOIN #{this} other
+      ON #{this}.journable_id = other.journable_id AND
+         #{this}.journable_type = other.journable_type AND
+         #{this}.version = other.version")
+      .where("#{this}.id < other.id")
+      .select("#{this}.id id, other.id duplicate_id")
+      .uniq_by(&:id)
+      .map { |pair| [pair.id, pair.duplicate_id] }
+  end
+
+  def journals_equivalent?(a, b)
+    unless records_equivalent?(a, b)
+      say 'Difference found in table journals', subitem: true
+      return false
+    end
+
+    unless records_equivalent?(a.data, b.data)
+      say 'Difference found in related data table (e.g. work_package_journals)', subitem: true
+      return false
+    end
+
+    a_attachments = a.attachable_journals.pluck(:attachment_id).sort
+    b_attachments = b.attachable_journals.pluck(:attachment_id).sort
+    unless a_attachments == b_attachments
+      say 'Difference found in attachable_journals', subitem: true
+      return false
+    end
+
+    a_custom_fields = customizable_journals_to_hash a.customizable_journals
+    b_custom_fields = customizable_journals_to_hash b.customizable_journals
+    unless a_custom_fields == b_custom_fields
+      say 'Difference found in customizable_journals', subitem: true
+      return false
+    end
+
+    true
+  end
```

Comment:
This one could actually be broken down by moving each check group into it's own method. That would leave this method with being just a boolean expression.